packages/jest-cli/src/search_source.js

```diff
@@ -209,14 +209,22 @@ export default class SearchSource {
     } else if (globalConfig.findRelatedTests && paths && paths.length) {
       return Promise.resolve(this.findRelatedTestsFromPattern(paths));
     } else {
+      const allFiles = this._context.hasteFS.getAllFiles();
       const validTestPaths =
         paths &&
         paths.filter(name => {
           try {
-            return fs.lstatSync(name).isFile();
+            if (!fs.lstatSync(name).isFile()) {
+              // It exists, but it is not a file; return false.
+              return false;
+            }
           } catch (e) {
+            // It does not exist; return false.
             return false;
           }
+
+          // It exists and it is a file; return true if it's in the project.
+          return allFiles.includes(path.resolve(name));
```

Comment:
@mjesun have you measured perf impact of this? I suspect `allFiles` to be quite a big array, no?

Reply:
I can change it to a `Set` so that the search becomes O(1).

Reply:
I mean, for the typical use-case (`jest some-path` or `jest test1 test2`) it should be negligible, but hypothetically it might hit the user harder when they provide a glob with tens of paths.

I'm good with a set.